File: auto_permissions/providers/anthropic.py
"""Anthropic Claude REST API connector with support for Haiku 4.5 and 3.5."""
import json
import urllib.request
import urllib.error
from typing import Any, Dict, Optional

from .base import BaseProvider, parse_json_safely, resolve_api_key

class AnthropicProvider(BaseProvider):
    """Connects to Anthropic Messages API using standard urllib."""
    def __init__(
        self,
        api_key: Optional[str] = None,
        model: str = "claude-3-5-haiku-latest",
        temperature: float = 0.0,
        timeout: float = 4.5,
    ):
        self.model = model
        self.temperature = temperature
        self.timeout = timeout
        self.api_key = api_key or resolve_api_key("ANTHROPIC_API_KEY", "anthropic_api_key")

    def evaluate(self, system_prompt: str, prompt: str) -> Optional[Dict[str, Any]]:
        if not self.api_key:
            return None

        url = "https://api.anthropic.com/v1/messages"
        payload = {
            "model": self.model,
            "max_tokens": 1024,
            "system": system_prompt,
            "messages": [
                {"role": "user", "content": prompt}
            ],
            "temperature": self.temperature,
        }
        data = json.dumps(payload).encode("utf-8")
        headers = {
            "Content-Type": "application/json",
            "Accept": "application/json",
            "x-api-key": self.api_key,
            "anthropic-version": "2023-06-01",
        }
        req = urllib.request.Request(url, data=data, headers=headers)
        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                resp_data = json.loads(resp.read().decode("utf-8"))
                contents = resp_data.get("content", [])
                if contents:
                    text = contents[0].get("text", "")
                    return parse_json_safely(text)
        except Exception:
            return None
        return None

File: internal/providers/anthropic.go
package providers

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"
)

// AnthropicProvider connects to Anthropic Messages API.
type AnthropicProvider struct {
	APIKey      string
	Model       string
	Temperature float64
	Timeout     time.Duration
	MaxTokens   int
	client      *http.Client
}

func NewAnthropicProvider(apiKey, model string, temp float64, timeout time.Duration, maxTokens int) *AnthropicProvider {
	if apiKey == "" {
		apiKey = ResolveAPIKey("ANTHROPIC_API_KEY", "", "")
	}
	if model == "" || model == "auto" || model == "default" {
		model = "claude-3-5-haiku-latest"
	}
	if timeout <= 0 {
		timeout = 4500 * time.Millisecond
	}
	if maxTokens <= 0 {
		maxTokens = 1024
	}
	return &AnthropicProvider{
		APIKey:      apiKey,
		Model:       model,
		Temperature: temp,
		Timeout:     timeout,
		MaxTokens:   maxTokens,
		client:      &http.Client{Timeout: timeout},
	}
}

func (p *AnthropicProvider) GetEndpoint() string {
	return "https://api.anthropic.com"
}

func (p *AnthropicProvider) Evaluate(systemPrompt, prompt string) (map[string]interface{}, string, error) {
	if p.APIKey == "" {
		return nil, "CLOUD", fmt.Errorf("anthropic api key not configured")
	}

	url := "https://api.anthropic.com/v1/messages"
	payload := map[string]interface{}{
		"model":       p.Model,
		"max_tokens":  p.MaxTokens,
		"system":      systemPrompt,
		"temperature": p.Temperature,
		"messages": []map[string]string{
			{"role": "user", "content": prompt},
		},
	}

	data, err := json.Marshal(payload)
	if err != nil {
		return nil, "CLOUD", err
	}

	req, err := http.NewRequest("POST", url, bytes.NewReader(data))
	if err != nil {
		return nil, "CLOUD", err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "application/json")
	req.Header.Set("x-api-key", p.APIKey)
	req.Header.Set("anthropic-version", "2023-06-01")

	resp, err := p.client.Do(req)
	if err != nil {
		return nil, "CLOUD", err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, "CLOUD", fmt.Errorf("anthropic returned status %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, "CLOUD", err
	}

	var dataResp struct {
		Content []struct {
			Text string `json:"text"`
		} `json:"content"`
	}

	if err := json.Unmarshal(body, &dataResp); err == nil && len(dataResp.Content) > 0 {
		text := dataResp.Content[0].Text
		parsed := ParseJSONSafely(text)
		if parsed != nil {
			return parsed, "CLOUD", nil
		}
	}

	return nil, "CLOUD", fmt.Errorf("failed to parse anthropic response")
}
